// tsb/src/common/events.js
"use strict";
var CompositeDisposible = (function () {
    function CompositeDisposible() {
        this.disposibles = [];
    }
    CompositeDisposible.prototype.add = function (disposible) {
        this.disposibles.push(disposible);
    };
    CompositeDisposible.prototype.dispose = function () {
        this.disposibles.forEach(function (d) { return d.dispose(); });
    };
    return CompositeDisposible;
})();
exports.CompositeDisposible = CompositeDisposible;
var TypedEvent = (function () {
    function TypedEvent() {
        var _this = this;
        this.listeners = [];
        this.listenersOncer = [];
        this.on = function (listener) {
            _this.listeners.push(listener);
            return {
                dispose: function () { return _this.off(listener); }
            };
        };
        this.once = function (listener) {
            _this.listenersOncer.push(listener);
        };
        this.off = function (listener) {
            var callbackIndex = _this.listeners.indexOf(listener);
            if (callbackIndex > -1)
                _this.listeners.splice(callbackIndex, 1);
        };
        this.emit = function (event) {
            _this.listeners.forEach(function (listener) { return listener(event); });
            _this.listenersOncer.forEach(function (listener) { return listener(event); });
            _this.listenersOncer = [];
            _this._last = event;
            while (_this._currentQueue.length) {
                var item = _this._currentQueue.pop();
                item.resolve(event);
            }
        };
        this.pipe = function (te) {
            return _this.on(function (e) { return te.emit(e); });
        };
        this._currentQueue = [];
        this._last = null;
        this.current = function () {
            if (_this._last != null) {
                return Promise.resolve(_this._last);
            }
            else {
                var p = Promise.defer();
                _this._currentQueue.push(p);
                return p.promise;
            }
        };
    }
    return TypedEvent;
})();
exports.TypedEvent = TypedEvent;
var SingleListenerQueue = (function () {
    function SingleListenerQueue() {
        var _this = this;
        this.listener = null;
        this.pending = [];
        this.pendingCount = function () { return _this.pending.length; };
    }
    SingleListenerQueue.prototype.on = function (listener) {
        var _this = this;
        this.listener = listener;
        this.pending.forEach(function (evt) { return _this.listener(evt); });
        this.pending = [];
    };
    SingleListenerQueue.prototype.off = function () {
        this.listener = null;
    };
    SingleListenerQueue.prototype.emit = function (event) {
        if (!this.listener) {
            this.pending.push(event);
        }
        else {
            this.listener(event);
        }
    };
    return SingleListenerQueue;
})();
exports.SingleListenerQueue = SingleListenerQueue;
