// tsb/src/common/types.js
"use strict";
exports.cacheDir = './.tsb';
exports.title = "TypeScript Builder";
(function (TriState) {
    TriState[TriState["Unknown"] = 0] = "Unknown";
    TriState[TriState["True"] = 1] = "True";
    TriState[TriState["False"] = 2] = "False";
})(exports.TriState || (exports.TriState = {}));
var TriState = exports.TriState;
exports.errors = {
    CALLED_WHEN_NO_ACTIVE_PROJECT_FOR_FILE_PATH: "A query *that needs an active project* was made when there is no active project for given filePath"
};
function getRefactoringsByFilePath(refactorings) {
    var loc = {};
    for (var _i = 0, refactorings_1 = refactorings; _i < refactorings_1.length; _i++) {
        var refac = refactorings_1[_i];
        if (!loc[refac.filePath])
            loc[refac.filePath] = [];
        loc[refac.filePath].push(refac);
    }
    for (var filePath in loc) {
        var refactorings_2 = loc[filePath];
        refactorings_2.sort(function (a, b) {
            return (b.span.start - a.span.start);
        });
    }
    return loc;
}
exports.getRefactoringsByFilePath = getRefactoringsByFilePath;
(function (FilePathType) {
    FilePathType[FilePathType["File"] = 0] = "File";
    FilePathType[FilePathType["Dir"] = 1] = "Dir";
})(exports.FilePathType || (exports.FilePathType = {}));
var FilePathType = exports.FilePathType;
