// tsb/src/server/utils/fsu.js
"use strict";
function consistentPath(filePath) {
    return filePath.split('\\').join('/');
}
exports.consistentPath = consistentPath;
var path = require("path");
var fs = require("fs");
var mkdirp = require('mkdirp');
function resolve() {
    var args = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        args[_i - 0] = arguments[_i];
    }
    return consistentPath(path.resolve.apply(path, args));
}
exports.resolve = resolve;
function isExt(path, ext) {
    return path && ext && path.indexOf(ext, path.length - ext.length) !== -1;
}
exports.isExt = isExt;
function makeRelativePath(relativeFolder, filePath) {
    var relativePath = path.relative(relativeFolder, filePath).split('\\').join('/');
    if (relativePath[0] !== '.') {
        relativePath = './' + relativePath;
    }
    return relativePath;
}
exports.makeRelativePath = makeRelativePath;
function removeExt(filePath) {
    return filePath.substr(0, filePath.lastIndexOf('.'));
}
exports.removeExt = removeExt;
function readFile(filePath) {
    return fs.readFileSync(filePath, 'utf8');
}
exports.readFile = readFile;
function writeFile(filePath, content) {
    mkdirp.sync(path.dirname(filePath));
    fs.writeFileSync(filePath, content, 'utf8');
}
exports.writeFile = writeFile;
exports.existsSync = function (filePath) { return fs.existsSync(filePath); };
exports.isAbsolute = function (filePath) { return path.isAbsolute(filePath); };
exports.isDir = function (filePath) { return fs.lstatSync(filePath).isDirectory(); };
