// tsb/src/server/workers/fileListing/fileListingMaster.js
"use strict";
var sw = require("../../utils/simpleWorker");
var contract = require("./fileListingContract");
var events_1 = require("../../../common/events");
var workingDir = require("../../disk/workingDir");
exports.filePathsPartial = new events_1.TypedEvent();
exports.filePathsCompleted = new events_1.TypedEvent();
exports.fileChangedOnDisk = new events_1.TypedEvent();
exports.initialIndexComplete = false;
var Master;
(function (Master) {
    Master.increment = function (q) {
        return Promise.resolve({
            num: ++q.num
        });
    };
    Master.fileListUpdated = function (q) {
        if (q.completed) {
            exports.initialIndexComplete = true;
            exports.filePathsCompleted.emit({ filePaths: q.filePaths });
        }
        else {
            exports.initialIndexComplete = false;
            exports.filePathsPartial.emit({ filePaths: q.filePaths });
        }
        return Promise.resolve({});
    };
    Master.fileChanged = function (q) {
        exports.fileChangedOnDisk.emit({ filePath: q.filePath });
        return Promise.resolve({});
    };
})(Master || (Master = {}));
var _checkTypes = Master;
exports.worker = sw.startWorker(__dirname + '/fileListingWorker', contract.worker, Master).worker;
function start() {
    exports.worker.setupWatch({ directory: workingDir.getProjectRoot() });
}
exports.start = start;
