// tsb/src/socket/socketContract.js
"use strict";
var socketLib_1 = require("../socketLib/socketLib");
exports.server = {
    echo: {},
    filePaths: {},
    makeAbsolute: {},
    openFile: {},
    closeFile: {},
    editFile: {},
    saveFile: {},
    getFileStatus: {},
    availableProjects: {},
    getActiveProjectConfigDetails: {},
    setActiveProjectConfigDetails: {},
    isFilePathInActiveProject: {},
    setOpenUITabs: {},
    getOpenUITabs: {},
    getFilePathsInActiveProject: {},
    getErrors: {},
    getCompletionsAtPosition: {},
    quickInfo: {},
    getRenameInfo: {},
    getDefinitionsAtPosition: {},
    getReferences: {},
    getDoctorInfo: {},
    formatDocument: {},
    formatDocumentRange: {},
};
exports.client = {
    increment: {},
};
exports.cast = {
    hello: new socketLib_1.TypedEvent(),
    filePathsCompleted: new socketLib_1.TypedEvent(),
    filePathsPartial: new socketLib_1.TypedEvent(),
    savedFileChangedOnDisk: new socketLib_1.TypedEvent(),
    didEdit: new socketLib_1.TypedEvent(),
    didStatusChange: new socketLib_1.TypedEvent(),
    errorsUpdated: new socketLib_1.TypedEvent(),
    availableProjectsUpdated: new socketLib_1.TypedEvent(),
    activeProjectConfigDetailsUpdated: new socketLib_1.TypedEvent(),
};
