// tsb/src/app/activeLists.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var utils = require("../common/utils");
var styles = require("./styles/styles");
var React = require("react");
var csx = require('csx');
var ui_1 = require("./ui");
var ui = require("./ui");
var commands = require("./commands/commands");
var clipboard_1 = require("./clipboard");
var icon_1 = require("./icon");
var DraggableCore = ui.DraggableCore;
var react_redux_1 = require("react-redux");
var state_1 = require("./state/state");
var state = require("./state/state");
var gotoHistory = require("./gotoHistory");
var notificationKeyboardStyle = {
    border: '2px solid',
    borderRadius: '6px',
    display: 'inline-block',
    padding: '5px',
    background: 'grey',
};
var resizerWidth = 5;
var resizerStyle = {
    background: 'radial-gradient(#444,transparent)',
    height: resizerWidth + 'px',
    cursor: 'ns-resize',
    color: '#666',
};
var ActiveLists = (function (_super) {
    __extends(ActiveLists, _super);
    function ActiveLists(props) {
        var _this = this;
        _super.call(this, props);
        this.toggleErrors = function () {
            if (_this.props.errorsExpanded) {
                state_1.collapseErrors({});
            }
            else {
                state_1.expandErrors({});
            }
        };
        this.openErrorLocation = function (error) {
            gotoHistory.gotoError(error);
        };
        this.openFile = function (filePath) {
            commands.doOpenOrFocusFile.emit({ filePath: filePath });
        };
        this.handleDrag = function (evt, ui) {
            _this.setState({ height: utils.rangeLimited({ num: _this.state.height - ui.position.deltaY, min: 100, max: window.innerHeight - 100 }) });
        };
        this.handleStop = function () {
        };
        this.state = {
            height: 150,
        };
    }
    ActiveLists.prototype.componentDidMount = function () {
        this.disposible.add(commands.toggleMessagePanel.on(function () {
            state.getState().errorsExpanded ? state.collapseErrors({}) : state.expandErrors({});
        }));
    };
    ActiveLists.prototype.render = function () {
        var _this = this;
        var errorCount = utils.selectMany(Object.keys(this.props.errorsByFilePath).map(function (k) { return _this.props.errorsByFilePath[k]; })).length;
        var errorPanel = undefined;
        if (this.props.errorsExpanded) {
            errorPanel = React.createElement("div", null, React.createElement(DraggableCore, {onDrag: this.handleDrag, onStop: this.handleStop}, React.createElement("div", {style: csx.extend(csx.flexRoot, csx.centerCenter, resizerStyle)}, React.createElement(icon_1.Icon, {name: "ellipsis-h"}))), React.createElement("div", {style: csx.extend(styles.errorsPanel.main, { height: this.state.height })}, errorCount ?
                Object.keys(this.props.errorsByFilePath)
                    .filter(function (filePath) { return !!_this.props.errorsByFilePath[filePath].length; })
                    .map(function (filePath, i) {
                    var errors = _this.props.errorsByFilePath[filePath]
                        .map(function (e, j) { return (React.createElement("div", {key: i + ":" + j, style: csx.extend(styles.hand, styles.errorsPanel.errorDetailsContainer), onClick: function () { return _this.openErrorLocation(e); }}, React.createElement("div", {style: styles.errorsPanel.errorDetailsContent}, React.createElement("div", {style: styles.errorsPanel.errorMessage}, "🐛(", e.from.line + 1, ":", e.from.ch + 1, ") ", e.message, ' ', React.createElement(clipboard_1.Clipboard, {text: e.filePath + ":" + (e.from.line + 1) + " " + e.message})), e.preview ? React.createElement("div", {style: styles.errorsPanel.errorPreview}, e.preview) : ''))); });
                    return React.createElement("div", {key: i}, React.createElement("div", {style: styles.errorsPanel.filePath, onClick: function () { return _this.openErrorLocation(_this.props.errorsByFilePath[filePath][0]); }}, React.createElement(icon_1.Icon, {name: "file-code-o", style: { fontSize: '.8rem' }}), " ", filePath), React.createElement("div", {style: styles.errorsPanel.perFileList}, errors));
                }) : React.createElement("div", {style: styles.errorsPanel.success}, "No Errors ❤")));
        }
        return (React.createElement("div", null, React.createElement(ui.VelocityTransitionGroup, {enter: { animation: "slideDown", duration: 150 }, leave: { animation: "slideUp", duration: 150 }}, errorPanel)));
    };
    ActiveLists = __decorate([
        react_redux_1.connect(function (state) {
            return {
                errorsExpanded: state.errorsExpanded,
                activeProject: state.activeProject,
                activeProjectFiles: state.activeProjectFilePathTruthTable,
                currentFilePath: state.currentFilePath,
                errorsByFilePath: state.errorsByFilePath,
                socketConnected: state.socketConnected
            };
        }), 
        __metadata('design:paramtypes', [Object])
    ], ActiveLists);
    return ActiveLists;
})(ui_1.BaseComponent);
exports.ActiveLists = ActiveLists;
