// tsb/src/app/clipboard.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var React = require("react");
var ui_1 = require("./ui");
var ui = require("./ui");
var buttonStyle = {
    height: '18px',
    padding: '2px 3px',
    display: 'inline-block',
    cursor: 'pointer',
    backgroundImage: 'linear-gradient(#7B7B7B, #353434)',
    border: '1px solid #464646',
    borderRadius: '3px',
    userSelect: 'none',
    outline: '0',
    ':active': {
        backgroundImage: 'linear-gradient(#353434, #7B7B7B)',
    }
};
var clippy = {
    width: '12px',
    height: '10px'
};
var Clipboard = (function (_super) {
    __extends(Clipboard, _super);
    function Clipboard() {
        _super.apply(this, arguments);
    }
    Clipboard.prototype.render = function () {
        return (React.createElement("button", {style: buttonStyle, "data-clipboard-text": this.props.text, onClick: function (event) { return event.stopPropagation(); }}, React.createElement("img", {src: "assets/clippy.svg", style: clippy})));
    };
    Clipboard = __decorate([
        ui.Radium, 
        __metadata('design:paramtypes', [])
    ], Clipboard);
    return Clipboard;
})(ui_1.BaseComponent);
exports.Clipboard = Clipboard;
