// tsb/src/app/pendingRequestsIndicator.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var React = require("react");
var ui_1 = require("./ui");
var react_redux_1 = require("react-redux");
require('./pendingRequestsIndicator.css');
var PendingRequestsIndicator = (function (_super) {
    __extends(PendingRequestsIndicator, _super);
    function PendingRequestsIndicator() {
        _super.apply(this, arguments);
    }
    PendingRequestsIndicator.prototype.componentWillReceiveProps = function () {
    };
    PendingRequestsIndicator.prototype.render = function () {
        var _this = this;
        var pendingRequestsCount = this.props.pendingRequests.length;
        var style = {
            opacity: pendingRequestsCount ? 1 : 0,
            transition: 'opacity .5s',
            cursor: 'pointer'
        };
        var pendingRequestCountStyle = {
            fontSize: '7px',
            paddingRight: '2px',
            color: '#2776b7'
        };
        return (React.createElement("span", {style: style, onClick: function () { return console.log(_this.props.pendingRequests); }}, React.createElement("span", {style: pendingRequestCountStyle}, pendingRequestsCount), React.createElement("span", {className: "loader"}, React.createElement("ul", null, React.createElement("li", {className: "one"}), React.createElement("li", {className: "two"}), React.createElement("li", {className: "three"}), React.createElement("li", {className: "four"}), React.createElement("li", {className: "five"})))));
    };
    PendingRequestsIndicator = __decorate([
        react_redux_1.connect(function (state) {
            return {
                pendingRequests: state.pendingRequests,
            };
        }), 
        __metadata('design:paramtypes', [])
    ], PendingRequestsIndicator);
    return PendingRequestsIndicator;
})(ui_1.BaseComponent);
exports.PendingRequestsIndicator = PendingRequestsIndicator;
