// tsb/src/app/styles/styles.js
"use strict";
exports.errorColor = '#f92672';
exports.successColor = '#00c990';
exports.tabHeaderContainer = {
    background: 'grey'
};
exports.tabHeader = {
    fontFamily: 'Roboto, sans-serif',
    fontWeight: 'bold',
    paddingLeft: '10px',
    paddingRight: '10px',
    paddingTop: '5px',
    paddingBottom: '5px',
    background: "rgb(66, 66, 66)",
    color: "rgb(150,150,150)",
    borderLeft: '6px solid rgb(88, 88, 88)',
    cursor: 'pointer',
    userSelect: 'none',
    transition: "color .5s, border-left .5s, background .5s"
};
exports.tabHeaderActive = {
    color: "#6edeef",
    borderLeft: '6px solid #6edeef',
};
exports.tabHeaderUnsaved = {
    background: "#777",
};
exports.fullWindow = {
    position: 'absolute',
    left: '10px',
    right: '10px',
    top: '10px',
    bottom: '10px',
};
exports.padded1 = {
    padding: '.25rem'
};
exports.padded1TopBottom = {
    paddingTop: exports.padded1.padding,
    paddingBottom: exports.padded1.padding
};
exports.padded1LeftRightBottom = {
    paddingLeft: exports.padded1.padding,
    paddingRight: exports.padded1.padding,
    paddingBottom: exports.padded1.padding
};
exports.padded2 = {
    padding: '.5rem'
};
exports.fullSize = {
    width: '100%',
    height: '100%',
};
var statusBarHeight = 25;
exports.statusBar = {
    height: statusBarHeight + "px",
    color: '#999',
    paddingLeft: '3px',
    paddingRight: '3px',
    fontSize: '12px',
    border: '1px solid #999',
};
exports.statusBarSection = {
    paddingLeft: '3px',
    paddingRight: '3px',
};
exports.statusBarError = {
    color: exports.errorColor
};
exports.statusBarSuccess = {
    color: exports.successColor
};
exports.noSelect = {
    userSelect: 'none',
    cursor: 'default'
};
exports.hand = {
    cursor: 'pointer',
};
exports.codeFont = {
    fontFamily: 'monospace'
};
var errorsPanel;
(function (errorsPanel) {
    errorsPanel.success = {
        color: exports.successColor
    };
    errorsPanel.main = {
        color: '#999',
        fontFamily: exports.codeFont.fontFamily,
        padding: '6px',
        overflow: 'auto',
    };
    errorsPanel.filePath = {
        fontSize: '1rem',
        fontWeight: 'bold',
        padding: '8px 8px 8px 0px',
        cursor: exports.hand.cursor,
    };
    errorsPanel.perFileList = {
        paddingLeft: '6px',
        borderLeft: '6px solid ' + exports.errorColor
    };
    errorsPanel.errorDetailsContainer = {
        padding: '3px',
    };
    errorsPanel.errorDetailsContent = {
        padding: '3px',
    };
    errorsPanel.errorMessage = {
        paddingBottom: '3px',
        cursor: 'pointer',
        userSelect: 'none',
    };
    errorsPanel.errorPreview = {
        padding: '3px',
        background: 'black',
        border: '2px solid #999',
        cursor: 'pointer',
    };
})(errorsPanel = exports.errorsPanel || (exports.errorsPanel = {}));
var modal;
(function (modal) {
    modal.keyStrokeStyle = {
        fontSize: '0.9rem',
        background: '#111',
        paddingTop: '1px',
        paddingBottom: '1px',
        paddingLeft: '4px',
        paddingRight: '4px',
        borderRadius: '5px',
        border: '2px solid'
    };
    modal.inputStyle = {
        backgroundColor: 'rgb(42,42,42)',
        color: 'white',
        outline: 'none',
        padding: '2px',
        fontSize: '1.5rem',
        lineHeight: '2rem',
        fontFamily: 'monospace',
        border: '3px solid #3C3C3C',
        transition: 'border .2s',
        ':focus': {
            boxShadow: '0px 0px 1px 1px #3C3C3C'
        }
    };
    modal.previewContainerStyle = {
        background: exports.tabHeader.background,
        padding: '5px',
    };
})(modal = exports.modal || (exports.modal = {}));
exports.hiddenInput = {
    height: '0px',
    width: '0px',
    color: 'transparent',
    background: 'transparent',
    position: 'absolute',
    top: '0px'
};
