// tsb/src/server/commandLine.js
"use strict";
var minimist = require('minimist');
var utils = require("../common/utils");
var workingDir = require("./disk/workingDir");
var fsu = require("./utils/fsu");
var defaultPort = 4444;
var argv = minimist(process.argv.slice(2), {
    string: ['dir'],
    boolean: ['open', 'safe'],
    alias: {
        'p': 'port',
        'd': 'dir',
        'o': 'open',
    },
    default: {
        p: defaultPort,
        d: process.cwd(),
        o: true
    }
});
exports.getOptions = utils.once(function () {
    var options = {
        port: argv.p,
        dir: argv.d,
        open: argv.o,
        safe: argv.safe,
        filePaths: [],
    };
    if (typeof options.port !== 'number') {
        options.port = defaultPort;
    }
    if (argv.d) {
        options.dir = workingDir.makeAbsoluteIfNeeded(argv.d);
        workingDir.setProjectRoot(options.dir);
    }
    if (argv._ && argv._.length) {
        options.filePaths = argv._.map(function (x) { return workingDir.makeAbsoluteIfNeeded(x); });
    }
    if (options.filePaths.length == 1) {
        var filePath = workingDir.makeAbsoluteIfNeeded(options.filePaths[0]);
        if (fsu.isDir(filePath)) {
            workingDir.setProjectRoot(filePath);
            options.filePaths = [];
        }
    }
    if (options.safe) {
        console.log("---SAFE MODE---");
    }
    return options;
});
